tests/cpp/test_multidevice_gpu_comms.cpp

```diff
@@ -0,0 +1,195 @@
+  CUDA_CALL(cudaFree(d_ptr));
+}
+
+TEST_F(GpuCommTest, IpcMemHandlePtrArithmeticAtReceiver) {
+  // TLDR; We can do pointer arithmetic on the receiver side.
+
+  // Allocate GPU memory
+  constexpr size_t size = 2 * sizeof(int64_t);
+  const int64_t num_devices = communicator_->size();
+  const int64_t rank = communicator_->deviceId();
+  const int64_t peer_rank = (rank + 1) % num_devices;
+  void* d_ptr;
+  CUDA_CALL(cudaMalloc(&d_ptr, size));
+
+  std::vector<int64_t> values;
+  values.push_back(2 * rank);
+  values.push_back(2 * rank + 1);
+  CUDA_CALL(cudaMemcpy(d_ptr, values.data(), size, cudaMemcpyHostToDevice));
+
+  cudaIpcMemHandle_t ipc_handle;
+  CUDA_CALL(cudaIpcGetMemHandle(&ipc_handle, d_ptr));
+
+  auto store = communicator_->getTcpStore();
+  store->set("ipc_handle_" + std::to_string(rank), toBytes(ipc_handle));
+  communicator_->barrier();
+  auto peer_ipc_handle = fromBytes<cudaIpcMemHandle_t>(
+      store->get("ipc_handle_" + std::to_string(peer_rank)));
+
+  int64_t* peer_d_ptr;
+  CUDA_CALL(cudaIpcOpenMemHandle(
+      (void**)&peer_d_ptr, peer_ipc_handle, cudaIpcMemLazyEnablePeerAccess));
+
+  int64_t peer_value;
+  CUDA_CALL(cudaMemcpy(
+      &peer_value, peer_d_ptr + 1, size / 2, cudaMemcpyDeviceToHost));
+
+  EXPECT_EQ(2 * peer_rank + 1, peer_value);
+
+  // Clean up
+  CUDA_CALL(cudaIpcCloseMemHandle(peer_d_ptr));
+  CUDA_CALL(cudaFree(d_ptr));
+}
+
+TEST_F(GpuCommTest, IpcMemHandlePtrArithmeticAtSender) {
+  // TLDR; We CANNOT do pointer arithmetic on the sender side! The IPC handle
+  // points to the beginning of the allocated buffer.
+
+  // Allocate GPU memory
+  constexpr size_t size = 2 * sizeof(int64_t);
+  const int64_t num_devices = communicator_->size();
+  const int64_t rank = communicator_->deviceId();
+  const int64_t peer_rank = (rank + 1) % num_devices;
+  int64_t* d_ptr;
+  CUDA_CALL(cudaMalloc(&d_ptr, size));
+
+  std::vector<int64_t> values;
+  values.push_back(2 * rank);
+  values.push_back(2 * rank + 1);
+  CUDA_CALL(cudaMemcpy(d_ptr, values.data(), size, cudaMemcpyHostToDevice));
+
+  cudaIpcMemHandle_t ipc_handle;
+  CUDA_CALL(cudaIpcGetMemHandle(&ipc_handle, d_ptr + 1));
+
+  auto store = communicator_->getTcpStore();
+  store->set("ipc_handle_" + std::to_string(rank), toBytes(ipc_handle));
+  communicator_->barrier();
+  auto peer_ipc_handle = fromBytes<cudaIpcMemHandle_t>(
+      store->get("ipc_handle_" + std::to_string(peer_rank)));
+
+  int64_t* peer_d_ptr;
+  CUDA_CALL(cudaIpcOpenMemHandle(
+      (void**)&peer_d_ptr, peer_ipc_handle, cudaIpcMemLazyEnablePeerAccess));
+
+  int64_t peer_value;
+  CUDA_CALL(
+      cudaMemcpy(&peer_value, peer_d_ptr, size / 2, cudaMemcpyDeviceToHost));
+
+  EXPECT_EQ(
+      2 * peer_rank,
+      peer_value); // and not 2 * peer_rank + 1 as could be expected!
+
+  // Clean up
+  CUDA_CALL(cudaIpcCloseMemHandle(peer_d_ptr));
+  CUDA_CALL(cudaFree(d_ptr));
+}
+
+class StreamOpTest : public NVFuserTest {};
+
+TEST_F(StreamOpTest, StreamWriteValue32) {
```

Comment:
What's this test for? I.e. how is cuStreamWriteValue32 going to be used for the higher level p2ps or collectives?

Reply:
`cuStreamWriteValue32` and `cuStreamWaitValue32` are used for synchronizations between devices, see https://github.com/NVIDIA/Fuser/pull/3911